internal/api/get_credentials_v1alpha1.go

```diff
@@ -0,0 +1,88 @@
+package api
+
+import (
+	"context"
+
+	"connectrpc.com/connect"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	kubeerr "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	typesv1alpha1 "github.com/akuity/kargo/internal/api/types/v1alpha1"
+	svcv1alpha1 "github.com/akuity/kargo/pkg/api/service/v1alpha1"
+)
+
+func (s *server) GetCredentials(
+	ctx context.Context,
+	req *connect.Request[svcv1alpha1.GetCredentialsRequest],
+) (*connect.Response[svcv1alpha1.GetCredentialsResponse], error) {
+	project := req.Msg.GetProject()
+	if err := validateFieldNotEmpty("project", project); err != nil {
+		return nil, err
+	}
+
+	name := req.Msg.GetName()
+	if err := validateFieldNotEmpty("name", name); err != nil {
+		return nil, err
+	}
+
+	if err := s.validateProjectExists(ctx, project); err != nil {
+		return nil, err
+	}
+
+	secret := corev1.Secret{}
+	if err := s.client.Get(
+		ctx,
+		types.NamespacedName{
+			Namespace: project,
+			Name:      name,
+		},
+		&secret,
+	); err != nil {
+		if kubeerr.IsNotFound(err) {
+			return nil, connect.NewError(connect.CodeNotFound, err)
+		}
+		return nil, errors.Wrap(err, "get secret")
+	}
+
+	// If this isn't labeled as repository credentials, return not found.
+	var isCredentials bool
+	if secret.Labels != nil {
+		_, isCredentials = secret.Labels[kargoapi.CredentialTypeLabelKey]
+	}
+	if !isCredentials {
+		return nil, connect.NewError(
+			connect.CodeNotFound,
+			errors.Errorf("secret %q exists, but is not labeled as credentials", name),
+		)
+	}
+
+	secret = redactCredentialSecretValues(secret)
+
+	return connect.NewResponse(&svcv1alpha1.GetCredentialsResponse{
+		Credentials: typesv1alpha1.ToSecretProto(&secret),
+	}), nil
+}
+
+// redactCredentialSecretValues returns a copy of the secret with all values in
+// the stringData map redacted except for those with specific keys that are
+// known to represent non-sensitive information when used correctly. Note: The
+// primary intention, at present, is only to redact the value associated with
+// the "password" key, but this approach prevents accidental exposure of the
+// password in the event that it has accidentally been assigned to a
+// wrong/unknown key, such as "pass" or "passwd".
+func redactCredentialSecretValues(secret corev1.Secret) corev1.Secret {
+	secret.StringData = make(map[string]string, len(secret.Data))
+	for k, v := range secret.Data {
+		switch k {
+		case "repoURL", "repoURLPattern", "username":
+			secret.StringData[k] = string(v)
+		default:
+			secret.StringData[k] = "*** REDACTED ***"
+		}
+	}
+	secret.Data = nil
+	return secret
+}
```

Comment:
There was one corner case that bit us in Argo CD. When the Secret was applied via Kubectl directly to Kube API server, we forgot that the sensitive data was additionally included in the `last-applied-configuration` annotation. And so the secret data could leak through that annotation and added logic to strip that specific annotation.

Next, another user reported another config management tool (I forget which), had the equivalent of `last-applied-configuration` (which Argo CD did not know about), and so we had to remove that.

That said, labels, given their character constraints, will likely never be used the same way as annotations and contain sensitive data.

Reply:
> `last-applied-configuration`

Nice though @jessesuen 

I can blank that out.

Reply:
> another config management tool (I forget which), had the equivalent of last-applied-configuration

There may not be much reason to not just blank out all annotations when retrieving creds in this manner.

Reply:
@jessesuen see 56be70b48e97f619c2182785c5834d68bed5514b

Reply:
Is there user value in redacting the annotation field values? Or would simply stripping all annotations be a simpler and less noisy approach.

Reply:
(Another config management tool having the equivalent of last-applied-configuration is KPT Config Sync by the way, using `configmanagement.gke.io/declared-config`).

Reply:
But this gets me thinking now about what is the _right_ way for someone to update creds.

The counterpart to `kargo get creds` is `kargo update creds`. The update endpoint carries out an update operation... I should change it to a patch.

Changing this to draft for a bit...

Reply:
Updated in 47be6461fa8a0e451d2b9f1bbf4f03baa68fc7e9

It's still an update (not a patch), but it now makes specific changes to a secret instead of overwriting it wholesale.